### DISCUSSION TITLE
Another option in video player

### DISCUSSION THREAD

**Discussion by lupin3rd** [brunoherbelin/vimix#154] [Ideas]
It would be very useful to add another mode of behavior when the player reaches the end of the video.
There are currently 3:
- stop at end
- loop to start
- bounce
In many theater show i need to deactivate the video and "stop at end" is good only if the last frame of the video is black.
I think that we need also a "deactivate at end" that deactivate the source.

**Comment by lupin3rd** [comment]
What do you think about @brunoherbelin ?

**Reply by brunoherbelin** [reply]
Yes, good idea!
I am implementing it.

It was possible in GLmixer, and was dropped for simplicity. There was another option possible : rewind and stop, so that it loops, shows the first frame of the video and stops. Not sure this would be useful...

**Reply by lupin3rd** [reply]
In some very particular cases it could be useful...
"deactivate at end" it is much more frequent, because very often you want to show a video only once and deactivate it, now I add black frames at the bottom of the video :smile: 

Thank you.

**Comment by brunoherbelin** [comment]
New play option 'Stop and blackout at end' is now implemented. 
You can checkout the beta or release.

NB: i didn't do the rewind because I want vimix to have only the minimum amount of useful features, and this one seemed useless.

**Reply by lupin3rd** [reply]
I test last beta. Works fine!
Why the source go to alpha 0 but don't go out (deactivate)?
If i would to reactivate the same source i need to send an OSC messago alpha/0 and then an alpha/1 ?

**Reply by brunoherbelin** [reply]
From the programing side, the detection of end of timeline is not done at the level of control of the sources (it is inside the timeline), so changing the alpha requires another mechanism for the timeline to inform the source, which is more heavy to implement and this is why i did it like that at first.

Then, I thought of still doing this and moving the source out of the ring to deactivate it when reacing the end: but then there is another problem, more tricky. Here is why : when users set a session with a video visible, and set it to deactivate at end, they expect that at next time they load the session, the video will open, play and deactivate at end. BUT : if they save the session after it has been automatically deactivated at end, the session stores the status of the video to be placed at alpha = 0 (or even outside mixing ring), and when loading the session next time, they will wonder why the video is not playing at all !

This problem is now handled in the program by a special state of 'auto-stopped-state' of the timeline which is separate to the video being saved as being playing. But the program cannot know if an action of moving the source out of the ring was automatic or caused by the user. So until I find a specific mechanism for that, I prefer to avoid this feature.

**Reply by lupin3rd** [reply]
Thanks for the clear explanation.
The way you implemented the functionality seems like a fair compromise to me.
The only doubt I had was whether the timeline blackout respected the layer order.
In that case if the next video was below in the layer, I would see a black box.
But I saw that the order is not respected and therefore it is OK. In practice the video becomes transparent even if present.
Thanks again for this feature!
